### DISCUSSION TITLE
Parse error with iCloud calendar subscription

### DISCUSSION THREAD

**Discussion by TheArkive** [bitfireAT/icsx5#150] [Q&A] [answered]
I have a few observations to report (and fix ideas) regarding iCloud public calendar subscriptions.

Your site mentions that this is supported, and at one time maybe it was.  I just tried it, and here's how it went:

- I input the `webcal://` link
- the `webcal://` prefix was automatically changed to http
- error message saying http isn't supported, must use https
- i changed the link to https
- error: Can't parse iCal file

I can replicate this process and post a full error message in an issue if necessary.

The current public URL for iCloud calendars does not end in `.ics`, at least not in this case.  If I change the `webcal://` prefix to `http://` or `https://` and past this link into a browser, then I can actually download the public calendar link as an ICS file.

Is it possible to use these mechanics in this app to create a read-only subscription sync setup in this app?

Here's the headers used in my browser (Firefox) to do the download (i've censored some info, cuz I'm not trying to broadcast the actual calendar:

```
HTTP/1.1 200 OK
Server: AppleHttpServer/xxxx
Date: Mon, 01 May 2023 14:46:07 GMT
Content-Type: text/calendar; charset=UTF-8
Transfer-Encoding: chunked
Connection: keep-alive
X-Responding-Instance: caldavj:3:p124-caldavj-xxxx-xxxx:xxxx:xxxx:xxxx
X-Apple-API-Version: v1
ETag: "xxxx=="
content-encoding: gzip
Strict-Transport-Security: max-age=31536000; includeSubDomains;
x-apple-user-partition: 124
via: xrail:icloud-xrail-group81-ext-xxxx-xxxx:xxxx:xxxx:xxxx,xxxx:xxxx:xxxx
X-Apple-Request-UUID: xxxx-xxxx-xxxx-xxxx-xxxx
access-control-expose-headers: X-Apple-Request-UUID,Via
X-Apple-Edge-Response-Time: 43
```

**Comment by sunkup** [comment]
Hi!
The fact that you get the last error "Can't parse iCal file" should not have anything to do with the used protocol. We can take a look at the iCal file, to try and find out why parsing fails.

Can you provide us with the ics file, which you can download via the browser? You may XX out any private information again.

**Comment by TheArkive** [comment]
I would certainly love to send it!

Is there an email I could send this to?

This ics file is 24,400 lines long.  Anonymizing it completely would be difficult.

**Reply by sunkup** [reply]
Right, you can send it to [support@davx5.com](mailto:support@davx5.com) or use the [support form](https://www.davx5.com/support) on our website.

**Reply by TheArkive** [reply]
Very cool, I'll do that now.  Thanks.

**Reply by rfc2822** [reply, marked as answer]
Thanks for the file. It contains this invalid event:

```
BEGIN:VEVENT
…
SUMMARY:Office Dinner
X-APPLE-STRUCTURED-LOCATION;VALUE=URI;X-ADDRESS="xxxxxxxxxxxxxx x, xxxxx
 xxxxxxxxxxxxxxxxxx, xxxxxxx";X-APPLE-MAPKIT-HANDLE=xxxxxxxxxxxxxxxxxxxxx
 ;X-TITLE=Facility \"Some Fancy Name\":geo:11.111111,22.222222
END:VEVENT
END:VCALENDAR
```

The problem is the `X-TITLE=Facility \"Some Fancy Name\"` parameter. [RFC 5545 3.2](https://www.rfc-editor.org/rfc/rfc5545#section-3.2):

> Property parameter values MUST NOT contain the DQUOTE character.  The DQUOTE character is used as a delimiter for parameter values that contain restricted characters or URI text.

For exactly this case – to allow double-quotes in parameter values –, [RFC 6868](https://www.rfc-editor.org/rfc/rfc6868) has been created. With RFC 6868, the parameter would be `X-TITLE=Facility ^'Some Fancy Name^'`.

So the file is invalid and ical4j can't parse it, so ICSx5 can't parse it :/

BTW, there's no PRODID in the file, so we don't know who has generated the broken iCalendar :( Is it directly from iCloud or was it uploaded somewhere and maybe modified by some other software?

**Reply by TheArkive** [reply]
Thanks very much for finding that.  After I deleted it, now the parsing works fine.

This file came directly from iCloud.  I sent myself the URL after making the calendar public, then changed "webcal" to "https" and pasted it into a web browser, and downloaded it.

As for other software modifying it, I'll have to ask my wife exactly how she made the event, and hopefully she will remember.  As far as I know, she only makes calendar events on her phone.

I suppose it's possible that someone else sent out an invalid ical file, and Apple made it easy to simply import the event.

If I ever find out that different software was used to generate this event, I'll post it here.

Thanks again 🎉.  Now my wife and I can finally share each other's calendars.
